x-pack/legacy/plugins/infra/public/pages/logs/analysis/sections/analyze_in_ml_button.tsx

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import url from 'url';
+import { EuiButton } from '@elastic/eui';
+import { FormattedMessage } from '@kbn/i18n/react';
+import chrome from 'ui/chrome';
+import { encode } from 'rison-node';
+import { TimeRange } from '../../../../../common/http_api/shared/time_range';
+
+export const AnalyzeInMlButton: React.FunctionComponent<{
+  jobId: string;
+  partition?: string;
+  timeRange: TimeRange;
+}> = ({ jobId, partition, timeRange }) => {
+  const pathname = chrome.addBasePath('/app/ml');
+  const buttonLabel = (
+    <FormattedMessage
+      id="xpack.infra.logs.analysis.analyzeInMlButtonLabel"
+      defaultMessage="Analyze in ML"
+    />
+  );
+  return partition ? (
+    <EuiButton
+      fill={false}
+      href={getPartitionSpecificSingleMetricViewerLink(pathname, jobId, partition, timeRange)}
+    >
+      {buttonLabel}
+    </EuiButton>
+  ) : (
+    <EuiButton fill={true} href={getOverallAnomalyExplorerLink(pathname, jobId, timeRange)}>
+      {buttonLabel}
+    </EuiButton>
+  );
+};
+
+const getOverallAnomalyExplorerLink = (pathname: string, jobId: string, timeRange: TimeRange) => {
+  const { from, to } = convertTimeRangeToParams(timeRange);
+
+  const _g = encode({
+    ml: {
+      jobIds: [jobId],
+    },
+    time: {
+      from,
+      to,
+    },
+  });
+
+  const hash = `/explorer?_g=${_g}`;
+
+  return url.format({
+    pathname,
+    hash,
+  });
+};
+
+const getPartitionSpecificSingleMetricViewerLink = (
+  pathname: string,
+  jobId: string,
+  partition: string,
+  timeRange: TimeRange
+) => {
+  const { from, to } = convertTimeRangeToParams(timeRange);
+
+  const _g = encode({
+    ml: {
+      jobIds: [jobId],
+    },
+    time: {
+      from,
+      to,
+      mode: 'absolute',
+    },
+  });
+
+  const _a = encode({
+    mlTimeSeriesExplorer: {
+      entities: { 'event.dataset': partition },
+    },
+  });
+
+  const hash = `/timeseriesexplorer?_g=${_g}&_a=${encodeURIComponent(_a)}`;
```

Comment:
From experience there are situations where the all-mighty angular router chooses to disagree with the standard url encoding. It might be safer to use the `QueryString.encode` function from `ui/utils/query_string` to avoid that. Not sure if that is likely to happen here, but `encodeURIComponent` rings an alarm bell with me. :wink:

So it could look something like

```typescript
const hash = `/timeseriesexplorer?${QueryString.encode({_g, _a})}`;
```

Reply:
Thank you for the hint, changed 👍 There seems to be quite a lot of discrepancy across all of the different Kibana plugins when it comes to links (understandable to a degree with React vs Angular, but even then even within the same routing environment).